### ISSUE TITLE
Debug as part of algorithm trait? Or in util?

### ISSUE THREAD

**Issue by mhoward2718** [p2t2/figaro#406]
Scala offers an annotation called `@elidable`, which tells the compiler to not to include calls to annotated methods in the resulting bytecode if a flag is set. We could unify our debugging statements by creating a function which accepts a string and just prints it to a log or stream, and use this annotation on it. We could add the function to the algorithm trait, or perhaps as globally accessible in util.

Could doing this this help our code coverage goal? I am not sure how scoverage deals with methods that have this annotation, or if adding the scoverage ignore annotation to the debug function would ignore calls throughout the code base. From a micro-optimization perspective, this would also remove all the branches which result from checking `if (debug)` before printing.

**Comment by apfeffer** [comment]
Cool!

+  +  +  +  +  +  +  +  +  +  +  +
Avi Pfeffer
Charles River Analytics Inc.
apfeffer@cra.com<mailto:apfeffer@cra.com>
(617) 491-3474x513

**Comment by gtakata** [comment]
Coverage is measured by tests unit and otherwise, while your debug statements are in the main code, so the annotations won’t help with this. However, they would help to provide switchable debugging in the main code itself.

Can you also use it to switch off compiling? In C++ you can identify sections of code (through the header) and turn on/off compiling those sections depending on an external flag.

*********************
Glenn Takata
Charles River Analytics Inc.
617.491.3474 x625
www.cra.com<http://www.cra.com>

**Comment by bruttenberg** [comment]
Yeah, this could potentially be useful, but I’d like more details on how it’s implemented. For example, is the switch at compile time or run time? Right now we can do runtime debugging switches, but if its compile time, then when we distribute the jar, people wouldn’t be able to run the debug code (unless we distributed the debug jars).

+  +  +  +  +  +  +  +  +  +  +  +  +  +
Brian Ruttenberg
Charles River Analytics Inc.
617.491.3474 x730
www.cra.com<http://www.cra.com/>
